<a href="https://colab.research.google.com/github/blacknotgrey/SF_Data_science/blob/main/Blok_1_Python_10_Pandas_csv.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ЗАПИСЬ В CSV-ФАЙЛ
Предположим, что мы захотели сохранить созданный нами ранее DataFrame. Самым простым и распространённым источником табличных данных является формат csv (comma-separated values). В данном формате ячейки таблицы обозначаются некоторым разделителем, чаще всего запятой либо точкой с запятой.
Экспорт данных в формат csv осуществляется с помощью метода DataFrame to_csv()

Основные параметры метода DataFrame to_csv()
path_or_buf — путь до файла, в который будет записан DataFrame (например, data/my_data.csv);
sep — разделитель данных в выходном файле (по умолчанию ',');
decimal — разделитель чисел на целую и дробную части в выходном файле (по умолчанию '.');
columns — список столбцов, которые нужно записать в файл (по умолчанию записываются все столбцы);
index — параметр, определяющий, требуется ли создавать дополнительный столбец с индексами строк в файле (по умолчанию True).

Заранее создадим папку data в директории, где лежит наш ноутбук. Теперь давайте сохраним наш DataFrame с информацией о странах в csv-файл countries.csv и положим файл в папку data. При этом укажем, что разделителем в нашем файле будет являться символ ';', а также то, что нам не нужен дополнительный столбец с индексами строк

In [1]:
import pandas as pd

In [2]:
melb_data = pd.read_csv('data/melb_data.csv', sep=',')

In [3]:
display(melb_data)

index         Suburb           Address  Rooms Type      Price Method  \
0          0     Abbotsford      85 Turner St      2    h  1480000.0      S   
1          1     Abbotsford   25 Bloomburg St      2    h  1035000.0      S   
2          2     Abbotsford      5 Charles St      3    h  1465000.0     SP   
3          3     Abbotsford  40 Federation La      3    h   850000.0     PI   
4          4     Abbotsford       55a Park St      4    h  1600000.0     VB   
...      ...            ...               ...    ...  ...        ...    ...   
13575  13575  Wheelers Hill      12 Strada Cr      4    h  1245000.0      S   
13576  13576   Williamstown     77 Merrett Dr      3    h  1031000.0     SP   
13577  13577   Williamstown       83 Power St      3    h  1170000.0      S   
13578  13578   Williamstown      96 Verdon St      4    h  2500000.0     PI   
13579  13579     Yarraville        6 Agnes St      4    h  1285000.0     SP   

        SellerG        Date  Distance  ...  Car  Landsize  BuildingArea  \
0        Biggin   3/12/2016       2.5  ...  1.0     202.0         126.0   
1        Biggin   4/02/2016       2.5  ...  0.0     156.0          79.0   
2        Biggin   4/03/2017       2.5  ...  0.0     134.0         150.0   
3        Biggin   4/03/2017       2.5  ...  1.0      94.0         126.0   
4        Nelson   4/06/2016       2.5  ...  2.0     120.0         142.0   
...         ...         ...       ...  ...  ...       ...           ...   
13575     Barry  26/08/2017      16.7  ...  2.0     652.0         126.0   
13576  Williams  26/08/2017       6.8  ...  2.0     333.0         133.0   
13577     Raine  26/08/2017       6.8  ...  4.0     436.0         126.0   
13578   Sweeney  26/08/2017       6.8  ...  5.0     866.0         157.0   
13579   Village  26/08/2017       6.3  ...  1.0     362.0         112.0   

       YearBuilt  CouncilArea  Lattitude  Longtitude  \
0         1970.0        Yarra  -37.79960   144.99840   
1         1900.0        Yarra  -37.80790   144.99340   
2         1900.0        Yarra  -37.80930   144.99440   
3         1970.0        Yarra  -37.79690   144.99690   
4         2014.0        Yarra  -37.80720   144.99410   
...          ...          ...        ...         ...   
13575     1981.0          NaN  -37.90562   145.16761   
13576     1995.0          NaN  -37.85927   144.87904   
13577     1997.0          NaN  -37.85274   144.88738   
13578     1920.0          NaN  -37.85908   144.89299   
13579     1920.0          NaN  -37.81188   144.88449   

                       Regionname  Propertycount           Coordinates  
0           Northern Metropolitan         4019.0    -37.7996, 144.9984  
1           Northern Metropolitan         4019.0    -37.8079, 144.9934  
2           Northern Metropolitan         4019.0    -37.8093, 144.9944  
3           Northern Metropolitan         4019.0    -37.7969, 144.9969  
4           Northern Metropolitan         4019.0    -37.8072, 144.9941  
...                           ...            ...                   ...  
13575  South-Eastern Metropolitan         7392.0  -37.90562, 145.16761  
13576        Western Metropolitan         6380.0  -37.85927, 144.87904  
13577        Western Metropolitan         6380.0  -37.85274, 144.88738  
13578        Western Metropolitan         6380.0  -37.85908, 144.89299  
13579        Western Metropolitan         6543.0  -37.81188, 144.88449  

[13580 rows x 23 columns]

Какова цена объекта недвижимости под индексом 15?

In [4]:
melb_data.loc[15, 'Price']

1310000.0

Когда был продан объект под индексом 90?

In [5]:
melb_data.loc[90, 'Date']

'10/09/2016'

Во сколько раз площадь прилегающей территории, на которой находится здание с индексом 3521, больше площади участка, на котором находится здание с индексом 1690? Ответ округлите до целого числа

In [6]:
print(round(melb_data.loc[3521, 'Landsize'] / melb_data.loc[1690, 'Landsize']))

3


In [7]:
print(melb_data.loc[3521, 'Landsize'])
print(melb_data.loc[1690, 'BuildingArea'])
print(melb_data.loc[1690, 'Landsize'])

234.0
148.0
84.0


Исследование структуры DataFrame. Для этого у DataFrame есть методы head() и tail(), которые возвращают n первых и n последних строк таблицы соответственно (по умолчанию n = 5).

In [8]:
# Выведем первые пять строк и семь последних строк нашей таблицы
display(melb_data.head())
melb_data.tail(7)

index      Suburb           Address  Rooms Type      Price Method SellerG  \
0      0  Abbotsford      85 Turner St      2    h  1480000.0      S  Biggin   
1      1  Abbotsford   25 Bloomburg St      2    h  1035000.0      S  Biggin   
2      2  Abbotsford      5 Charles St      3    h  1465000.0     SP  Biggin   
3      3  Abbotsford  40 Federation La      3    h   850000.0     PI  Biggin   
4      4  Abbotsford       55a Park St      4    h  1600000.0     VB  Nelson   

        Date  Distance  ...  Car  Landsize  BuildingArea  YearBuilt  \
0  3/12/2016       2.5  ...  1.0     202.0         126.0     1970.0   
1  4/02/2016       2.5  ...  0.0     156.0          79.0     1900.0   
2  4/03/2017       2.5  ...  0.0     134.0         150.0     1900.0   
3  4/03/2017       2.5  ...  1.0      94.0         126.0     1970.0   
4  4/06/2016       2.5  ...  2.0     120.0         142.0     2014.0   

   CouncilArea  Lattitude  Longtitude             Regionname  Propertycount  \
0        Yarra   -37.7996    144.9984  Northern Metropolitan         4019.0   
1        Yarra   -37.8079    144.9934  Northern Metropolitan         4019.0   
2        Yarra   -37.8093    144.9944  Northern Metropolitan         4019.0   
3        Yarra   -37.7969    144.9969  Northern Metropolitan         4019.0   
4        Yarra   -37.8072    144.9941  Northern Metropolitan         4019.0   

          Coordinates  
0  -37.7996, 144.9984  
1  -37.8079, 144.9934  
2  -37.8093, 144.9944  
3  -37.7969, 144.9969  
4  -37.8072, 144.9941  

[5 rows x 23 columns]

index         Suburb        Address  Rooms Type      Price Method  \
13573  13573       Werribee    5 Nuragi Ct      4    h   635000.0      S   
13574  13574    Westmeadows     9 Black St      3    h   582000.0      S   
13575  13575  Wheelers Hill   12 Strada Cr      4    h  1245000.0      S   
13576  13576   Williamstown  77 Merrett Dr      3    h  1031000.0     SP   
13577  13577   Williamstown    83 Power St      3    h  1170000.0      S   
13578  13578   Williamstown   96 Verdon St      4    h  2500000.0     PI   
13579  13579     Yarraville     6 Agnes St      4    h  1285000.0     SP   

             SellerG        Date  Distance  ...  Car  Landsize  BuildingArea  \
13573  hockingstuart  26/08/2017      14.7  ...  1.0     662.0         172.0   
13574            Red  26/08/2017      16.5  ...  2.0     256.0         126.0   
13575          Barry  26/08/2017      16.7  ...  2.0     652.0         126.0   
13576       Williams  26/08/2017       6.8  ...  2.0     333.0         133.0   
13577          Raine  26/08/2017       6.8  ...  4.0     436.0         126.0   
13578        Sweeney  26/08/2017       6.8  ...  5.0     866.0         157.0   
13579        Village  26/08/2017       6.3  ...  1.0     362.0         112.0   

       YearBuilt  CouncilArea  Lattitude  Longtitude  \
13573     1980.0          NaN  -37.89327   144.64789   
13574     1970.0          NaN  -37.67917   144.89390   
13575     1981.0          NaN  -37.90562   145.16761   
13576     1995.0          NaN  -37.85927   144.87904   
13577     1997.0          NaN  -37.85274   144.88738   
13578     1920.0          NaN  -37.85908   144.89299   
13579     1920.0          NaN  -37.81188   144.88449   

                       Regionname  Propertycount           Coordinates  
13573        Western Metropolitan        16166.0  -37.89327, 144.64789  
13574       Northern Metropolitan         2474.0   -37.67917, 144.8939  
13575  South-Eastern Metropolitan         7392.0  -37.90562, 145.16761  
13576        Western Metropolitan         6380.0  -37.85927, 144.87904  
13577        Western Metropolitan         6380.0  -37.85274, 144.88738  
13578        Western Metropolitan         6380.0  -37.85908, 144.89299  
13579        Western Metropolitan         6543.0  -37.81188, 144.88449  

[7 rows x 23 columns]

Далее хотелось бы узнать размер таблицы — количество строк и количество столбцов. Это можно сделать с помощью атрибута shape, который возвращает кортеж с количеством строк и столбцов

In [9]:
melb_data.shape

(13580, 23)

Для того чтобы получить более детальную информацию о столбцах таблицы, можно использовать метод DataFrame info()

In [10]:
melb_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 23 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   index          13580 non-null  int64  
 1   Suburb         13580 non-null  object 
 2   Address        13580 non-null  object 
 3   Rooms          13580 non-null  int64  
 4   Type           13580 non-null  object 
 5   Price          13580 non-null  float64
 6   Method         13580 non-null  object 
 7   SellerG        13580 non-null  object 
 8   Date           13580 non-null  object 
 9   Distance       13580 non-null  float64
 10  Postcode       13580 non-null  int64  
 11  Bedroom        13580 non-null  float64
 12  Bathroom       13580 non-null  float64
 13  Car            13580 non-null  float64
 14  Landsize       13580 non-null  float64
 15  BuildingArea   13580 non-null  float64
 16  YearBuilt      13580 non-null  float64
 17  CouncilArea    12211 non-null  object 
 18  Lattit

Из вывода метода info() становится понятно, что в нашей таблице есть столбец CouncilArea с пропущенными значениями — количество непустых значений в столбце меньше, чем количество строк в таблице (12211 < 13580).
Пустыми, или пропущенными, значениями называются значения в ячейках таблицы, которые не заполнены по какой-либо причине, то есть на их месте стоит пустое место. В Pandas такие значения обозначаются символом NaN (Not-a-Number).
Как работать с "нанами" см. здесь: https://pythonru.com/biblioteki/not-a-number-vse-o-nan-pd-5

ИЗМЕНЕНИЕ ТИПА ДАННЫХ В СТОЛБЦЕ
Если присмотреться внимательнее к выводу метода info(), а конкретнее — к типам данных столбцов, становится понятно, что некоторые признаки кодируются не совсем корректными типами данных. 
Например, данные в столбцах, которые отражают количество, должны, по идее, выражаться целым числом (Car, Bedroom, Bathroom и Propertyсount), однако кодируются float64 — числом с плавающей запятой размером 64 бита.
Наконец, данные в столбце с годом постройки (YearBuilt) также представлены в формате чисел с плавающей точкой.

Чтобы исправить это, можно воспользоваться методом astype(), который позволяет преобразовать тип данных столбца

In [11]:
melb_data['Car'] = melb_data['Car'].astype('int64')
melb_data['Bedroom'] = melb_data['Bedroom'].astype('int64')
melb_data['Bathroom'] = melb_data['Bathroom'].astype('int64')
melb_data['Propertycount'] = melb_data['Propertycount'].astype('int64')
melb_data['YearBuilt'] = melb_data['YearBuilt'].astype('int64')
melb_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 23 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   index          13580 non-null  int64  
 1   Suburb         13580 non-null  object 
 2   Address        13580 non-null  object 
 3   Rooms          13580 non-null  int64  
 4   Type           13580 non-null  object 
 5   Price          13580 non-null  float64
 6   Method         13580 non-null  object 
 7   SellerG        13580 non-null  object 
 8   Date           13580 non-null  object 
 9   Distance       13580 non-null  float64
 10  Postcode       13580 non-null  int64  
 11  Bedroom        13580 non-null  int64  
 12  Bathroom       13580 non-null  int64  
 13  Car            13580 non-null  int64  
 14  Landsize       13580 non-null  float64
 15  BuildingArea   13580 non-null  float64
 16  YearBuilt      13580 non-null  int64  
 17  CouncilArea    12211 non-null  object 
 18  Lattit

ПОЛУЧЕНИЕ ОПИСАТЕЛЬНОЙ СТАТИСТИКИ

Часто при работе с таблицей нужно быстро посмотреть на основные статистические свойства её столбцов. Для этого можно воспользоваться методом DataFrame describe().

По умолчанию метод работает с числовыми (int64 и float64) столбцами и показывает число непустых значений (count), среднее (mean), стандартное отклонение (std), минимальное значение (min),  квантили уровней 0.25, 0.5 (медиана) и 0.75 (25%, 50%, 75%) и максимальное значение (max) для каждого столбца исходной таблицы.

Дополнительно:
https://ru.wikipedia.org/wiki/Среднеквадратическое_отклонение
https://rus.cubecartskins.net/quartile

In [12]:
# Чтобы не увязнуть в обилии информации, выведем на экран значение статистических параметров
# только для столбцов Distance (расстояние от объекта недвижимости до центра Мельбурна), BuildingArea (площадь здания) 
# и Price (цена объекта)

melb_data.describe().loc[:, ['Distance', 'BuildingArea' , 'Price']]

Distance  BuildingArea         Price
count  13580.000000  13580.000000  1.358000e+04
mean      10.137776    139.633972  1.075684e+06
std        5.868725    392.217403  6.393107e+05
min        0.000000      0.000000  8.500000e+04
25%        6.100000    122.000000  6.500000e+05
50%        9.200000    126.000000  9.030000e+05
75%       13.000000    129.940000  1.330000e+06
max       48.100000  44515.000000  9.000000e+06

Какие интересные выводы мы можем сделать, исходя из построенной таблицы?

Столбец Distance (расстояние до центра города): В нашем наборе данных представлены объекты, удалённые от центра города на расстояние до 48 километров.

Столбец BuildingArea (площадь здания): В нашей таблице представлены участки, на которых отсутствует само здание (его площадь равна 0), то есть владельцы продавали участок земли без строений на нём.
Большая часть объектов (75%) имеет площадь здания до 130 квадратных метров, однако в нашей таблице есть настоящий дворец (судя по его размерам) с площадью в 44 515 квадратных метров.

Столбец Price (цена): Диапазон цен на недвижимость в Мельбурне варьируется от 85 тысяч до 9 миллионов австралийских долларов.

На самом деле метод describe() можно применять не только к числовым признакам. С помощью параметра include можно указать тип данных, для которого нужно вывести описательную информацию.

Например, для типа данных object метод describe() возвращает DataFrame, в котором указаны:

количество непустых строк (count);
количество уникальных значений (unique);
самое частое значение — мода —  (top);
частота — объём использования — этого значения (freq) для каждого столбца типа object исходной таблицы

In [13]:
melb_data.describe(include=['object'])

Suburb           Address   Type Method SellerG        Date  \
count       13580             13580  13580  13580   13580       13580   
unique        314             13378      3      5     268          58   
top     Reservoir  36 Aberfeldie St      h      S  Nelson  27/05/2017   
freq          359                 3   9449   9022    1565         473   

       CouncilArea             Regionname         Coordinates  
count        12211                  13580               13580  
unique          33                      8               13097  
top       Moreland  Southern Metropolitan  -37.8361, 144.9966  
freq          1163                   4695                  12

Также приведём несколько выводов, которые можно сделать из полученной таблицы.

Столбец Suburb (пригород): Наибольшее количество проданных объектов (359) находилось в пригороде Reservoir.

Столбец Type (тип здания): Самый популярный тип дома — h (дом, вилла, коттедж).

Столбец SellerG (риелтор): В наших данных нам известно о 268 различных риэлторских компаниях, однако самой главной «акулой» в этом бизнесе является компания Nelson — они продали 1 565 различных домов.

Столбец Date (дата продажи): В нашей таблице содержится информация за 58 дней, при этом наибольшее число продаж (473) пришлось на 27 мая 2017 года.

ПОЛУЧЕНИЕ ЧАСТОТЫ УНИКАЛЬНЫХ ЗНАЧЕНИЙ В СТОЛБЦЕ

Для того чтобы определить, сколько раз в столбце повторяется каждый из вариантов значений (т.е. найти частоту для каждого уникального знания), используется метод value_counts().

Данный метод возвращает объект Series, в котором в качестве индексов выступают уникальные категории столбца, а значениями — соответствующая им частота.

In [14]:
# Рассмотрим работу value_counts() на примере столбца с названиями районов

melb_data['Regionname'].value_counts()

Southern Metropolitan         4695
Northern Metropolitan         3890
Western Metropolitan          2948
Eastern Metropolitan          1471
South-Eastern Metropolitan     450
Eastern Victoria                53
Northern Victoria               41
Western Victoria                32
Name: Regionname, dtype: int64

In [15]:
# Чтобы сделать вывод более интерпретируемым и понятным, 
# можно воспользоваться параметром normalize. При установке значения этого параметра на True 
# результат будет представляться в виде доли (относительной частоты)

melb_data['Regionname'].value_counts(normalize=True)

# Из результатов мы можем сделать вывод, что наименьшее количество объектов (менее 1%) 
# было продано в районах Victoria, а наибольшее — в районах Metropolitan.

Southern Metropolitan         0.345729
Northern Metropolitan         0.286451
Western Metropolitan          0.217084
Eastern Metropolitan          0.108321
South-Eastern Metropolitan    0.033137
Eastern Victoria              0.003903
Northern Victoria             0.003019
Western Victoria              0.002356
Name: Regionname, dtype: float64

In [16]:
melb_data['BuildingArea'].value_counts()

126.00    6492
120.00     114
110.00      89
100.00      88
130.00      84
          ... 
255.79       1
424.00       1
186.36       1
147.20       1
368.00       1
Name: BuildingArea, Length: 602, dtype: int64

In [17]:
melb_data['Type'].value_counts()

h    9449
u    3017
t    1114
Name: Type, dtype: int64

In [18]:
print(round(1114 / (9449 + 3017 + 1114) * 100))

8


АГРЕГИРУЮЩИЕ МЕТОДЫ

Агрегирующим в Pandas называется метод, который для каждого столбца возвращает только одно значение — показатель (например, вычисление медианы, максимума, среднего и так далее)

МЕТОД	СТАТИСТИЧЕСКИЙ ПАРАМЕТР
.count()	Количество непустых значений
.mean()	Среднее значение
.min()	Минимальное значение
.max()	Максимальное значение
.deviance()	Дисперсия
.std()	Стандартное отклонение
.sum()	Сумма
.quantile(x)	Квантиль уровня x
.nunique()	Число уникальных значений

Если один из этих методов применить ко всему DataFrame, то в результате его работы будет получен объект типа Series, в котором в качестве индексов будут выступать наименования столбцов, а в качестве значений — статистический показатель. В случае применения метода к отдельному столбцу результатом вычислений станет число.

В каждый метод можно передать некоторые параметры, среди которых:
axis  — определяет, подсчитывать параметр по строкам или по столбцам;
numeric_only — определяет, вычислять параметры только по числовым столбцам/строкам или нет (True/False).

In [19]:
# Вычислим среднюю цену на объекты недвижимости
print(melb_data['Price'].mean())

1075684.079455081


In [20]:
# Найдём максимальное количество парковочных мест
print(melb_data['Car'].max())

10


In [21]:
# А теперь представим, что риэлторская ставка для всех компаний за продажу недвижимости составляет 12%.
# Найдём общую прибыльность риэлторского бизнеса в Мельбурне. Результат округлим до сотых

rate = 0.12
income = melb_data['Price'].sum() * rate
print('Total income of real estate agencies:', round(income, 2))

Total income of real estate agencies: 1752934775.88


In [22]:
# Найдём, насколько медианная площадь территории отличается от её среднего значения. 
# Вычислим модуль разницы между медианой и средним
# и разделим результат на среднее, чтобы получить отклонение в долях

landsize_median = melb_data['Landsize'].median() 
landsize_mean =  melb_data['Landsize'].mean()
print(abs(landsize_median - landsize_mean)/landsize_mean)

# В результате получаем долю отклонения медианы от среднего значения.
# Умножив результат на 100, получим его в процентах.
# Отклонение медианы от среднего значения на 21% является довольно большим,
# и это повод задуматься над тем, чтобы исследовать признак на наличие аномалий.

0.21205713983546193


МОДАЛЬНОЕ ЗНАЧЕНИЕ

Отдельный интерес представляет статический показатель моды — самого распространённого значения в столбце. Он вычисляется с помощью метода mode(). Модальных значений может быть несколько, то есть несколько значений могут встречаться одинаковое количество раз. Поэтому метод mode(), в отличие от агрегирующих методов, возвращает не одно число, а серию.

In [23]:
# Вычислим, какое число комнат чаще всего представлено на рынке недвижимости

print(melb_data['Rooms'].mode())

0    3
Name: Rooms, dtype: int64


In [24]:
# Метод mode() может быть использован не только с числовыми столбцами, но и со столбцами типа object. 
# Так, например, с помощью следующего кода можно найти наиболее распространённое название района

print(melb_data['Regionname'].mode())

0    Southern Metropolitan
Name: Regionname, dtype: object


In [25]:
# Самый популярный тип объекта

print(melb_data['Type'].mode())

0    h
Name: Type, dtype: object


In [26]:
# Для каких столбцов метод std() выполнится с ошибкой?

print(melb_data.std())

index              3920.352663
Rooms                 0.955748
Price            639310.724296
Distance              5.868725
Postcode             90.676964
Bedroom               0.965921
Bathroom              0.691712
Car                   0.960793
Landsize           3990.669241
BuildingArea        392.217403
YearBuilt            29.088642
Lattitude             0.079260
Longtitude            0.103916
Propertycount      4378.581772
dtype: float64


/var/folders/x9/fgbykpz53tggdmyw4rxkp7mw0000gn/T/ipykernel_12184/2021779383.py:3: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  print(melb_data.std())


In [27]:
# Чему равно максимальное количество домов на продажу в районе

print(melb_data['Propertycount'].max())

21650


In [28]:
# Чему равно стандартное отклонение (разброс) расстояния от центра города до объекта недвижимости?

print(round(melb_data['Distance'].std()))

6


In [29]:
# Чему равно отклонение (в процентах) медианного значения площади здания от его среднего значения?

buildingArea_median = melb_data['BuildingArea'].median() 
buildingArea_mean =  melb_data['BuildingArea'].mean()
print(round((abs(buildingArea_median - buildingArea_mean)/buildingArea_mean) * 100))

10


In [30]:
# Задан ряд чисел [1, 2, 4, 2, 3, 2, 1, 5, 6]. Чему равна мода в данном ряду?

list_l = pd.Series([1, 2, 4, 2, 3, 2, 1, 5, 6])
list_mode = list_l.mode()

print(list_mode)

0    2
dtype: int64


In [31]:
# Сколько спален чаще всего встречается в домах в Мельбурне?

print(melb_data['Bedroom'].mode())

0    3
Name: Bedroom, dtype: int64


Фильтрация данных в DataFrame

Под фильтрацией в DataFrame подразумевается получение новой таблицы путём вырезания строк, не удовлетворяющих поставленному условию. 

Разберём классический способ фильтрации в DataFrame — фильтрацию с помощью масок.
Маской называется Series, которая состоит из булевых значений, при этом значения True соответствуют тем индексам, для которых заданное условие выполняется, в противном случае ставится значение False (например, цена > 2 млн).

In [32]:
# Создадим маску и положим её в переменную с именем mask

mask = melb_data['Price'] > 2000000
display(mask)

0        False
1        False
2        False
3        False
4        False
         ...  
13575    False
13576    False
13577    False
13578     True
13579    False
Name: Price, Length: 13580, dtype: bool

In [33]:
# Для фильтрации нужно просто подставить переменную mask в индексацию DataFrame. 
# Маска показывает, какие строки нужно оставлять в результирующем наборе, а какие — убирать 
# (выведем первые пять строк отфильтрованной таблицы)

display(melb_data[mask].head())

# В результате выполнения фильтрации возвращается новый DataFrame, 
# полученный из исходного, при этом исходная таблица melb_data остаётся без изменений.

index       Suburb               Address  Rooms Type      Price Method  \
80     80  Albert Park  112 Beaconsfield Pde      3    h  2850000.0     PI   
85     85  Albert Park     104 Richardson St      4    h  2300000.0      S   
88     88  Albert Park        29 Faussett St      2    h  2120000.0      S   
92     92  Albert Park           2 Dundas Pl      3    h  2615000.0      S   
93     93  Albert Park          23 Finlay St      5    h  2100000.0      S   

     SellerG        Date  Distance  ...  Car  Landsize  BuildingArea  \
80    Buxton   4/03/2017       3.3  ...    0     211.0         198.0   
85  Marshall   7/05/2016       3.3  ...    1     153.0         180.0   
88      Greg  10/09/2016       3.3  ...    1     199.0         107.0   
92    Cayzer  10/12/2016       3.3  ...    1     177.0         181.0   
93      Greg  10/12/2016       3.3  ...    1     237.0         126.0   

    YearBuilt   CouncilArea  Lattitude  Longtitude             Regionname  \
80       1890  Port Phillip   -37.8481    144.9499  Southern Metropolitan   
85       1880  Port Phillip   -37.8447    144.9523  Southern Metropolitan   
88       1900  Port Phillip   -37.8422    144.9554  Southern Metropolitan   
92       1880  Port Phillip   -37.8415    144.9585  Southern Metropolitan   
93       1970  Port Phillip   -37.8436    144.9557  Southern Metropolitan   

    Propertycount         Coordinates  
80           3280  -37.8481, 144.9499  
85           3280  -37.8447, 144.9523  
88           3280  -37.8422, 144.9554  
92           3280  -37.8415, 144.9585  
93           3280  -37.8436, 144.9557  

[5 rows x 23 columns]

In [34]:
# Также вовсе не обязательно заносить маску в отдельную переменную — 
# можно сразу вставлять условие в операцию индексации DataFrame, например:

melb_data[melb_data['Price'] > 2000000]

index          Suburb               Address  Rooms Type      Price  \
80        80     Albert Park  112 Beaconsfield Pde      3    h  2850000.0   
85        85     Albert Park     104 Richardson St      4    h  2300000.0   
88        88     Albert Park        29 Faussett St      2    h  2120000.0   
92        92     Albert Park           2 Dundas Pl      3    h  2615000.0   
93        93     Albert Park          23 Finlay St      5    h  2100000.0   
...      ...             ...                   ...    ...  ...        ...   
13521  13521  Port Melbourne          44 Garton St      4    t  2455000.0   
13523  13523         Prahran        69 Greville St      4    h  2668000.0   
13553  13553    Surrey Hills          20 Albert Cr      4    h  2720000.0   
13555  13555    Surrey Hills              3 Oak St      4    h  3100000.0   
13578  13578    Williamstown          96 Verdon St      4    h  2500000.0   

      Method   SellerG        Date  Distance  ...  Car  Landsize  \
80        PI    Buxton   4/03/2017       3.3  ...    0     211.0   
85         S  Marshall   7/05/2016       3.3  ...    1     153.0   
88         S      Greg  10/09/2016       3.3  ...    1     199.0   
92         S    Cayzer  10/12/2016       3.3  ...    1     177.0   
93         S      Greg  10/12/2016       3.3  ...    1     237.0   
...      ...       ...         ...       ...  ...  ...       ...   
13521     SP  Marshall  26/08/2017       3.5  ...    2     123.0   
13523      S    Biggin  26/08/2017       4.6  ...    2     383.0   
13553      S       Kay  26/08/2017      10.2  ...    2    1005.0   
13555     VB  Marshall  26/08/2017      10.2  ...    3     832.0   
13578     PI   Sweeney  26/08/2017       6.8  ...    5     866.0   

       BuildingArea  YearBuilt   CouncilArea  Lattitude  Longtitude  \
80            198.0       1890  Port Phillip  -37.84810   144.94990   
85            180.0       1880  Port Phillip  -37.84470   144.95230   
88            107.0       1900  Port Phillip  -37.84220   144.95540   
92            181.0       1880  Port Phillip  -37.84150   144.95850   
93            126.0       1970  Port Phillip  -37.84360   144.95570   
...             ...        ...           ...        ...         ...   
13521           0.0       2010           NaN  -37.83349   144.94840   
13523         126.0       1970           NaN  -37.84879   144.98882   
13553         126.0       1920           NaN  -37.82421   145.10352   
13555         126.0       1970           NaN  -37.83564   145.10919   
13578         157.0       1920           NaN  -37.85908   144.89299   

                  Regionname  Propertycount           Coordinates  
80     Southern Metropolitan           3280    -37.8481, 144.9499  
85     Southern Metropolitan           3280    -37.8447, 144.9523  
88     Southern Metropolitan           3280    -37.8422, 144.9554  
92     Southern Metropolitan           3280    -37.8415, 144.9585  
93     Southern Metropolitan           3280    -37.8436, 144.9557  
...                      ...            ...                   ...  
13521  Southern Metropolitan           8648   -37.83349, 144.9484  
13523  Southern Metropolitan           7717  -37.84879, 144.98882  
13553  Southern Metropolitan           5457  -37.82421, 145.10352  
13555  Southern Metropolitan           5457  -37.83564, 145.10919  
13578   Western Metropolitan           6380  -37.85908, 144.89299  

[1007 rows x 23 columns]